sasmodels/models/micromagnetic_FF_3D.c

```diff
@@ -0,0 +1,196 @@
+//Core-shell form factor for anisotropy field (Hkx, Hky and Hkz), Nuc and lonitudinal magnetization Mz
+static double
+form_volume(double nuc_radius, double nuc_thickness, double mag_radius, double mag_thickness, double hk_radius)
+{
+  if( nuc_radius+nuc_thickness>mag_radius+ mag_thickness)
+    return M_4PI_3 * cube(nuc_radius + nuc_thickness);
+  else
+    return M_4PI_3 * cube(mag_radius + mag_thickness);
+
+}
+
+
+
+static double fq(double q, double radius,
+ double thickness, double core_sld, double shell_sld, double solvent_sld)
+{
+ const double form = core_shell_fq(q,
+  radius,
+  thickness,
+  core_sld,
+  shell_sld,
+  solvent_sld);
+ return form;
+}
+
+
+static double reduced_field(double q, double Ms, double Hi,
+ double A)
+{
+  if( Hi > 1.0e-6 ) 
+    return Ms / (Hi + 2.0 * A * 4.0 * M_PI / Ms * q * q * 10.0);//q in 10e10 m-1, A in 10e-12 J/m, mu0 in 1e-7 
+  else 
+    return Ms / (1.0e-6 + 2.0 * A * 4.0 * M_PI / Ms * q * q * 10.0); 
+ }
+
+ static double DMI_length(double Ms, double D, double qval)
+ {
+   return 2.0 * D * 4.0 * M_PI / Ms / Ms * qval ; //q in 10e10 m-1, A in 10e-3 J/m^2, mu0 in 4 M_PI 1e-7 
+ }
+
+//Mz is defined as the longitudinal magnetisation component along the magnetic field.
+//In the approach to saturation this component is (almost) constant with magnetic 
+//field and simplfy reflects the nanoscale variations in the saturation magnetisation
+//value in the sample. The misalignment of the magnetisation due to perturbing
+//magnetic anisotropy or dipolar magnetic fields in the sample enter Mx and My,
+//the two transversal magnetisation components, reacting to a magnetic field.
+//The micromagnetic solution for the magnetisation are from Michels et al. PRB 94, 054424 (2016).
+
+static double fqMxreal( double x, double y, double z, double Mz, double Hkx, double Hky, double Hi, double Ms, double A, double D)
+{
+  const double q=MAG_VEC(x, y, z);  
+  const double f = reduced_field(q, Ms, Hi, A)*(Hkx*(1.0+reduced_field(q, Ms, Hi, A)*y*y/q/q)-Ms*Mz*x*z/q/q*(1.0+reduced_field(q, Ms, Hi, A)*DMI_length(Ms, D,q)*DMI_length(Ms, D,q))-Hky*reduced_field(q, Ms, Hi, A)*x*y/q/q)/(1.0+reduced_field(q, Ms, Hi, A)*(x*x+y*y)/q/q-square(reduced_field(q, Ms, Hi, A)*DMI_length(Ms, D,z)));
+  return f;
```

Comment:
Suggest `const double Hr = reduced_field(q, Ms, Hi, A); const double Dr = DMI_length(Ms, D, q);` since these are constant throughout the expression. It'll make the code easier to check.

Units on the final term `DMI_length(Ms, D, z)` don't match: z in Å rather than q in 1/Å.

Reply:
After revision:
```C99
static double fqMxreal( double qx, double qy, double qz, double Mz, double Hkx, double Hky, double Hi, double Ms, double A, double D)
{
  const double qsq = qx*qx + qy*qy + qz*qz;
  const double q = sqrt(q);
  const double Hr = reduced_field(q, Ms, Hi, A);
  const double f = Hr * (
    Hkx * (qsq + Hr*qy*qy)
    - Ms*Mz*qx*qz * (1.0 + Hr*square(DMI_length(Ms, D, q)))
    - Hky*Hr*qx*qy
    ) / (qsq + Hr*(qx*qx + qy*qy) - square(Hr*DMI_length(Ms, D, qz)*q));
  return f;
}

static double fqMximag(double qx, double qy, double qz, double Mz, double Hkx, double Hky, double Hi, double Ms, double A, double D)
{
  const double qsq = qx*qx + qy*qy + qz*qz;
  const double q = sqrt(q);
  const double Hr = reduced_field(q, Ms, Hi, A);
  const double f = -Hr*qsq*(
    Ms*Mz*(1.0 + Hr)*DMI_length(Ms, D, qy)
    + Hky*Hr*DMI_length(Ms, D, qz)
    ) / (qsq + Hr*(qx*qx + qy*qy) - square(Hr*DMI_length(Ms, D, qz)*q));
  return f;
}

static double fqMyreal( double qx, double qy, double qz, double Mz, double Hkx, double Hky, double Hi, double Ms, double A, double D)
{
  const double qsq = qx*qx + qy*qy + qz*qz;
  const double q = sqrt(q);
  const double Hr = reduced_field(q, Ms, Hi, A);
  const double f = Hr*(
    Hky*(qsq + Hr*x*x)
    - Ms*Mz*qy*qz*(1.0 + Hr*square(DMI_length(Ms, D, q)))
    - Hkx*Hr*qx*qy
    ) / (qsq + Hr*(qx*qx + qy*qy) - square(Hr*DMI_length(Ms, D, qz)*q));
  return f;
}

static double fqMyimag( double qx, double qy, double qz, double Mz, double Hkx, double Hky, double Hi, double Ms, double A, double D)
{
  const double qsq = qx*qx + qy*qy + qz*qz;
  const double q = sqrt(q);
  const double Hr = reduced_field(q, Ms, Hi, A);
  const double f = Hr*qsq*(
    Ms*Mz*(1.0 + Hr)*DMI_length(Ms, D, qx)
    - Hkx*Hr*DMI_length(Ms, D, qz)
    ) / (qsq + Hr*(qx*qx + qy*qy) - square(Hr*DMI_length(Ms, D, qz)*q));
  return f;
}
```

Note: I multiplied top and bottom by q² to avoid potential divide by zero at q=0.